### ISSUE TITLE
[BUG] Allocating tensor makes calling cute.jit a lot slower even it doesn't use the allocated tensor

### ISSUE THREAD

**Issue by kainzhong** [NVIDIA/cutlass#3275]
### Which component has the problem?

CuTe DSL

### Bug Report

**Describe the bug**
If I allocate a tensor by `torch.empty`, it makes calling a compiled cute.jit function slower even if that function doesn't do anything and doesn't use the allocated tensor

**Steps/Code to reproduce bug**
```
import time
import torch
from torch.cuda import nvtx

import cutlass
import cutlass.cute as cute

x_fake0 = cute.runtime.make_fake_compact_tensor(cute.BFloat16, (256, 4096), stride_order=(1, 0), memspace=cute.AddressSpace.gmem, assumed_align=16)
x_fake1 = cute.runtime.make_fake_compact_tensor(cute.BFloat16, (128, 2048), stride_order=(1, 0), memspace=cute.AddressSpace.gmem, assumed_align=16)
x_fake2 = cute.runtime.make_fake_compact_tensor(cute.BFloat16, (2048, 256), stride_order=(1, 0), memspace=cute.AddressSpace.gmem, assumed_align=16)
x_fake3 = cute.runtime.make_fake_compact_tensor(cute.BFloat16, (256, 128), stride_order=(1, 0), memspace=cute.AddressSpace.gmem, assumed_align=16)
x_fake4 = cute.runtime.make_fake_compact_tensor(cute.BFloat16, (128, 256), stride_order=(1, 0), memspace=cute.AddressSpace.gmem, assumed_align=16)
x_fake5 = cute.runtime.make_fake_compact_tensor(cute.BFloat16, (256, 1024), stride_order=(1, 0), memspace=cute.AddressSpace.gmem, assumed_align=16)
x_fake6 = cute.runtime.make_fake_compact_tensor(cute.BFloat16, (512, 2048), stride_order=(1, 0), memspace=cute.AddressSpace.gmem, assumed_align=16)
x_fake7 = cute.runtime.make_fake_compact_tensor(cute.BFloat16, (128, 4096), stride_order=(1, 0), memspace=cute.AddressSpace.gmem, assumed_align=16)
x_fake8 = cute.runtime.make_fake_compact_tensor(cute.BFloat16, (2048, 128), stride_order=(1, 0), memspace=cute.AddressSpace.gmem, assumed_align=16)
x0 = torch.empty((256, 4096), dtype=torch.bfloat16, device='cuda')
x1 = torch.empty((128, 2048), dtype=torch.bfloat16, device='cuda')
x2 = torch.empty((2048, 256), dtype=torch.bfloat16, device='cuda')
x3 = torch.empty((256, 128), dtype=torch.bfloat16, device='cuda')
x4 = torch.empty((128, 256), dtype=torch.bfloat16, device='cuda')
x5 = torch.empty((256, 1024), dtype=torch.bfloat16, device='cuda')
x6 = torch.empty((512, 2048), dtype=torch.bfloat16, device='cuda')
x7 = torch.empty((128, 4096), dtype=torch.bfloat16, device='cuda')
x8 = torch.empty((2048, 128), dtype=torch.bfloat16, device='cuda')

class NoLaunchKernel:

    @cute.jit
    def __call__(self, x0, x1, x2, x3, x4, x5, x6, x7, x8):
        return

class LaunchKernel:

    @cute.jit
    def __call__(self, x0, x1, x2, x3, x4, x5, x6, x7, x8):
        self.kernel(x0, x1, x2, x3, x4, x5, x6, x7, x8).launch(
            grid=(1,),
            block=(1,),
        )

    @cute.kernel
    def kernel(self, x0, x1, x2, x3, x4, x5, x6, x7, x8):
        return

WARMUP = 100
TRIALS = 10
ITERS = 400

def bench(kernel_cls, with_alloc):
    if kernel_cls is not None:
        compiled = cute.compile(kernel_cls(), x_fake0, x_fake1, x_fake2, x_fake3, x_fake4, x_fake5, x_fake6, x_fake7, x_fake8, options="--enable-tvm-ffi")
    else:
        compiled = lambda *args : None

    # Warmup (include the alloc path so caching allocator is hot)
    for i in range(WARMUP):
        if with_alloc:
            torch.empty(1, dtype=torch.uint8, device='cuda')
        compiled(x0, x1, x2, x3, x4, x5, x6, x7, x8)
    torch.cuda.synchronize()

    trial_us = []
    torch.cuda.cudart().cudaProfilerStart()
    for trial in range(TRIALS):
        t0 = time.perf_counter_ns()
        nvtx.range_push(f"Trial-{trial}")
        for i in range(ITERS):
            if with_alloc:
                torch.empty(1, dtype=torch.uint8, device='cuda')
            compiled(x0, x1, x2, x3, x4, x5, x6, x7, x8)
        torch.cuda.synchronize()
        nvtx.range_pop()
        t1 = time.perf_counter_ns()
        trial_us.append((t1 - t0) / ITERS / 1000)
    torch.cuda.cudart().cudaProfilerStop()

    mean = sum(trial_us) / len(trial_us)
    var = sum((x - mean) ** 2 for x in trial_us) / (len(trial_us) - 1)
    std = var ** 0.5
    suffix = '+Alloc' if with_alloc else ''
    name = "Nothing" if kernel_cls is None else kernel_cls.__name__
    print(f'RESULT {name}{suffix} {mean} {std}', flush=True)

bench(None, True)
bench(NoLaunchKernel, False)
bench(LaunchKernel,   False)
bench(NoLaunchKernel, True)
bench(LaunchKernel,   True)
```

**Expected behavior**
On GB200 this script outputs:
```
RESULT Nothing+Alloc 1.6487247500000002 0.11324282288056196
RESULT NoLaunchKernel 1.6442525 0.01118622689838812
RESULT LaunchKernel 3.37709725 0.048031188586312216
RESULT NoLaunchKernel+Alloc 5.747288 0.38881368030559027
RESULT LaunchKernel+Alloc 8.0602305 0.2851624457455979
```
"Nothing+Alloc" means it doesn't call cute.jit and only does allocation, so I assume allocation along takes 1.64us
Then I would suppose if I add this time to "NoLaunchKernel" and "LaunchKernel", it would be about the same number from "NoLaunchKernel+Alloc" and "LaunchKernel+Alloc" and I don't even pass the allocated empty tensor to my compiled function. However, the actual duration is a lot longer which doesn't seem to make sense to me. They should be about 1.64+1.64=3.28us, and 1.64+3.37=5.01us or even less since it doesn't use the allocated tensor so they should be able to overlap instead of run sequentially/

Worth to mention that I also tried this on an AGX Thor on dlcluster and that yields a reasonable result:
```
RESULT Nothing+Alloc 1.8182492499999998 0.029221814650331548
RESULT NoLaunchKernel 1.95853925 0.01587520056959913
RESULT LaunchKernel 4.24348425 0.0610531419708591
RESULT NoLaunchKernel+Alloc 3.7950477499999997 0.015231044660768708
RESULT LaunchKernel+Alloc 6.26364875 0.0763004097780063
```

**Environment details (please complete the following information):**
On NVIDIA's ptyche cluster
Using Docker container with image `gitlab-master.nvidia.com:5005/dl/transformerengine/transformerengine:main-pytorch-py3-devel-arm64`

**Additional context**

**Comment by brandon-yujie-sun** [comment]
@LongshengDu to help check, while you have been looking at the compilation time recently.
@kainzhong which DSL version are you using?

**Comment by kainzhong** [comment]
@brandon-yujie-sun 

```
➜ pip show nvidia_cutlass_dsl
Name: nvidia-cutlass-dsl
Version: 4.5.0
Summary: NVIDIA CUTLASS Python DSL
Home-page: 
Author: NVIDIA Corporation
Author-email: 
License: 
Location: /usr/local/lib/python3.12/dist-packages
Requires: nvidia-cutlass-dsl-libs-base
Required-by: 
```

Btw, personally I think it's not exactly a compilation issue since my timing doesn't include the compilation process. It only includes the duration to launch the compiled cute.jit function. I investigated a bit with Claude Code and it seems more likely to be an issue related to CUDA driver API calls (like something becomes blocking somehow) since the allocated tensor is not even used by the jit function. I don't know how to dig deeper though.

**Comment by brandon-yujie-sun** [comment]
Thanks @kainzhong I gave it a try with 4.5.0 wheel on a x86 machine with B100, but seems cannot repro what you observed. Do you see the same or is this arm specific?

**Comment by kainzhong** [comment]
Hi @brandon-yujie-sun 

I tried a few more machines (on dl-cluster) and here's my result:
B300, x64:
```
RESULT Nothing+Alloc 1.8713382500000002 0.02142232736636396                                                                                                                                                                                                                                                                                                                                                                                                                                  
RESULT NoLaunchKernel 3.1682005 0.07243035279954578                                                                                                                                                                                                                                                                                                                                                                                                                                          
RESULT LaunchKernel 5.10923875 0.033272641231406413                                                                                                                                                                                                                                                                                                                                                                                                                                          
RESULT NoLaunchKernel+Alloc 5.17299875 0.052211943493829044                                                                                                                                                                                                                                                                                                                                                                                                                                  
RESULT LaunchKernel+Alloc 7.3726775 0.028337115433845844                                                                                                                                                                                                                                                                                                                                                                                                                                     
```

AGX Thor, arm64:
```
RESULT Nothing+Alloc 1.8183205 0.03837027835230455
RESULT NoLaunchKernel 3.74126475 0.12293048388130447
RESULT LaunchKernel 6.05836275 0.30136567492013666
RESULT NoLaunchKernel+Alloc 5.689875750000001 0.0403975660888471
RESULT LaunchKernel+Alloc 8.28779125 0.17585403232530536
```

B100, x64:
```
RESULT Nothing+Alloc 1.9471370000000001 0.017872424709591023
RESULT NoLaunchKernel 3.28742925 0.6808910767639242
RESULT LaunchKernel 4.9141977500000005 0.03351332712622629
RESULT NoLaunchKernel+Alloc 5.0830494999999996 0.03706208224047872
RESULT LaunchKernel+Alloc 7.0920112500000005 0.03546747867276113
```

However when I try it on B200, arm64:
```
RESULT Nothing+Alloc 1.5610405 0.10084196447081606
RESULT NoLaunchKernel 3.2651685 0.49556025207755205
RESULT LaunchKernel 5.10492 0.048537717950979845
RESULT NoLaunchKernel+Alloc 7.4333089999999995 0.16328778298605204
RESULT LaunchKernel+Alloc 10.63778175 0.16505148655903476
```
which starts to look strange to me.

I don't think it's arm64 specific because that AGX Thor returned a kinda normal result. Could you please check B200+arm64 this specific combination?

**Comment by LongshengDu** [comment]
Hi, I did some digging on my local x64 B200 machine with older (4.5) and newer version cutlass-dsl (4.6), which had some improvement on function argument setup. Combined time is now close to the sum. I repeated it three times and only saw about 0.2–0.3 us extra.


```text
Nothing+Alloc         2.01 us
NoLaunchKernel        1.13 us
NoLaunchKernel+Alloc  3.34 us
LaunchKernel          3.48 us
LaunchKernel+Alloc    5.83 us
```

One important detail is that the `with_alloc` tensor from `torch.empty(1, ...)` is not saved or passed to the function. Python allocates and immediately frees this temporary tensor before calling the function with the existing x0 ... x8 tensors. TVM FFI still processes those 9 arguments even though `NoLaunchKernel` does nothing with them.

I tested these parts separately. Passing 9 raw torch tensor costs about 1.22 us, while passing pre-converted TVM-FFI tensors costs about 0.58 us. However, the `torch.empty()` allocate/free cost stayed around 2.15 us in both cases. This means TVM-FFI conversion has a cost, but it does not cause the large slowdown from the original result.

I also checked with nsys. torch uses a caching allocator: it requests a larger memory block from CUDA once, then reuses small pieces of that block. Destroying the temporary tensor returns its piece to torch's cache instead of returning it to CUDA. Because the warmup makes this cache hot, `torch.empty()` inside the timed loop does not need `cudaMalloc` or `cudaFree`. I also saw no event query or synchronization, so this does not look like a blocking CUDA driver call.

The alloca/free and function-call setup are CPU work on single thread, so they run in sequential order. The extra delay is most likely from host-side interaction between torch and TVM FFI argument setup. Both use CPU-side object allocation, so **running them together can cost more than adding their separate times**.

**Comment by kainzhong** [comment]
Hum I see. I guess it's true that it's some strange interaction of torch and tvm-ffi that makes such delay. Though I wonder why it's different on x86-64 and arm64. I also tried 4.6.0 and on arm64 it's still quite different from x86-64:
```
➜  cutlass git:(main) ✗ python tmp.py                                                                                                                                                                                                                                             
RESULT Nothing+Alloc 1.6861497499999998 0.1339583999297398                                                                                                                                                                                                                        
RESULT NoLaunchKernel 1.9783335000000002 0.03128568973544003                                                                                                                                                                                                                      
RESULT LaunchKernel 4.048531 0.07381639733525042                                                                                                                                                                                                                                  
RESULT NoLaunchKernel+Alloc 6.71964775 0.19321374240667175                                                                                                                                                                                                                        
RESULT LaunchKernel+Alloc 10.0315315 0.11066700754967593                                                                                                                                                                                                                          
➜  cutlass git:(main) ✗ arch                                                                                                                                                                                                                                                      
aarch64                                                                                                                                                                                                                                                                           
➜  cutlass git:(main) ✗ pip show nvidia_cutlass_dsl                                                                                                                                                                                                                               
Name: nvidia-cutlass-dsl                                                                                                                                                                                                                                                          
Version: 4.6.0                                                                                                                                                                                                                                                                    
Summary: NVIDIA CUTLASS Python DSL                                                                                                                                                                                                                                                
Home-page:                                                                                                                                                                                                                                                                        
Author: NVIDIA Corporation                                                                                                                                                                                                                                                        
Author-email:                                                                                                                                                                                                                                                                     
License:                                                                                                                                                                                                                                                                          
Location: /usr/local/lib/python3.12/dist-packages                                                                                                                                                                                                                                 
Requires: nvidia-cutlass-dsl-libs-base, nvidia-cutlass-dsl-libs-cu12                                                                                                                                                                                                              
Required-by:                                                                                                                                                                                                                                                                      
```

**Comment by LongshengDu** [comment]
It is quite possible down to in the different implementation for torch and tvm-ffi on x86-64 and arm64 systems, system library performance behavior can vary a bit across ABIs.

**Comment by kainzhong** [comment]
Thanks, I think that explains such behavior.
